### PR TITLE
Much better comment preservation

### LINKED ISSUE #2546: TypeScript removing comments where it shouldn't
comment.ts

```
if (false) {
    //I will be removed
}
//I will be removed
else {
    //I will not be removed
    var tmp;
    //I will be removed
}

try{
    //I will not be removed
    var tmp;
    //I will be removed
}
//I will be removed
catch(e){
    //I will not be removed
    var tmp;
    //I will be removed
}
```

comment.js

```
if (false) {
}
else {
    //I will not be removed
    var tmp;
}
try {
    //I will not be removed
    var tmp;
}
catch (e) {
    //I will not be removed
    var tmp;
}
```

Are there any additional flags that can be set for TS to keep _all_ comments?

[Same issue from CodePlex](https://typescript.codeplex.com/workitem/2573)

### LINKED ISSUE #6069: in 1.8 comments are removed from empty blocks and unreachable places regardless of the unset --removeComments flag
### Case 1

typescript

```
 try {
                // AB: might throw if a property is readonly
                data[key] = null;
            } catch(e) {
                // AB: Here we are swallowing the exception on the attempt to write to a read-only property.
                // An alternative for using try/catch would be to use Object.getOwnPropertyDescriptor
                // which, when called for every single enumerable property, is slow, compared to
                // a few failed attempts to write to read only.
            }
```

javascript

```
try {
                    // AB: might throw if a property is readonly
                    data[key] = null;
                }
                catch (e) {
                }
```
### Case 2

typescript

```
    try {
        stack = undefined;
        /* tslint:disable:no-throw */
        throw new Error();
        /* tslint:enable:no-throw */
    } catch (e) {
        stack = e.stack.split('\n');
    }
```

javascript

```
try {
            stack = undefined;
            /* tslint:disable:no-throw */
            throw new Error();
        }
        catch (e) {
            stack = e.stack.split('\n');
        }
```

### LINKED ISSUE #9873: No comment preservation for parenthesized expressions
@evmar ran into issues where generating a JSDoc comment for a type assertion for Closure did not get preserved.

Neither

``` ts
var x = (/** @type {Foo.Bar} */ y);
```

nor

``` ts
var x = /** @type {Foo.Bar} */ (y);
```

have their comments preserved, though I'm not sure which is more common.

### LINKED ISSUE #11755: Emit comment on catch clause
**TypeScript Version:**  nightly (2.1.0-dev.20161020)

**Code**

``` ts
try { }
catch (/*error*/ e) { }
```

**Expected output:**

Same as the input

**Actual output:** 

``` js
try { }
catch (e) { }
```

### LINKED ISSUE #17606: Keep leadingComments of CallExpression
**TypeScript Version:**  2.3.3 +

**Code**
```ts
export const a = /*#__PURE__*/create("a")
```

**Expected behavior:**
CallExpression.leadingComments should have comments

**Actual behavior:**
in https://astexplorer.net/
image

### PR DESCRIPTION
**Pull request by weswigham**
This clears out some large swaths of #17689 (which is now much more precise now that I've taken an initial look on what should/should not be functioning right now), specifically:
- [x] `YieldExpression`
- [x] `AwaitExpression`
- [x] `DeleteExpression`
- [x] `TypeOfExpression`
- [x] `VoidExpression`
- [x] `NewExpression`
- [x] `ElementAccessExpression`
- [x] `PropertyAccessExpression`
- [x] Block-level statements where the comments need to be considered attached to a parenthesis token (all of them)
- [x] `VariableDeclaration`
- [x] All import and export kinds

should now all preserve any comments of any kind between any of their tokens, regardless of if the token is actually in the AST, so long as the node is sufficiently similar to its original form.

Just a reminder: Comment preservation is important for tools based on TS (like our own quickfixes) which expect to be able to extract, transform, and insert arbitrary ranges without modification. You can actually see improvements in our own `organizeImports` baseline. 😄 

Things that might still not preserve comments correctly include any multitoken type nodes, and some TS-specific constructs (like namespaces and casts). I'm choosing to tackle those later (if at all) in a separate PR, since they need to be tested in a separate harness (as we have no transform-less emit option). Additionally, I am aware of a bug preventing multiline object literals from preserving comments after non-final commas (#10385). I'm leaving a fix for that out of this PR, since fixing it _may_ end up requiring reworking `emitList` a bit more than I want to when there's already a lot else going on.

I don't think I have an exhaustive list, but I know this:
Fixes #11755
Fixes #2546 (all cases mentioned in the thread)
Fixes #17606
Fixes #6069 probably (provided I understand the original issue correctly)
Fixes #9873 (Which is actually a collection of issues)

Side note on why the changeset is so big: The sourcemap changes should be from changing `emitTokenWithComment` to no longer add an extra sourcemap entry for the token we're emitting comments for. Baselines were captured when this was used for `return` statements, and while the baselines were technically "more precise" insofar as there were more mapping entries, typically the extra entries were for equal width spans (as a token always mapped to itself) making the extra entry rather pointless. That alone wouldn't usually be enough for me to care, but we actually have some places where the token _does_ map to more text (when you consider non-`return` tokens), and the sourcemaps for those locations were always inaccurate (as the transformers hadn't setup corrected ones). So in lieu of updating the transformers to add mappings for all these tokens, I'm just leaving them all mapped as they are today, and reverting `return` to how it was more simply mapped in the past. If it'd help, I can extract that change into a separate PR so this one's easier to review once it's merged.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
(Psst, @evmar)

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -2399,7 +2399,7 @@ namespace ts {
 
         function emitEnumMember(node: EnumMember) {
             emit(node.name);
-            emitInitializer(node.initializer);
+            emitInitializer(node.initializer, node.name.end + 1, node);
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: Why are these all plus one? Shouldn't the start position be the end of the name?

Comment (weswigham) [reply src/compiler/emitter.ts]: Yep. This one (and like one other?) were the first ones I wrote and for some reason thought that I needed to skip the length the token in question. Thanks for reminding me that I still hadn't gone back and changed them.

Comment (DanielRosenwasser) [reply src/compiler/emitter.ts]: Maybe that's a good indicator that we need tests for cases where trivia comes immediately after:

**Comment by rbuckton** [comment]
Have you run a benchmark for these changes against master? I want to be sure these changes don't cause a dramatic slowdown.

**Comment by weswigham** [comment]
image
I did an unscientific 10/10 comparison yesterday, actually. I'm not sure what our limit here is, but the difference doesn't _seem_ terribly outlandish.

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -2287,21 +2299,21 @@ namespace ts {
...
-            writeKeyword("default");
-            writePunctuation(":");
-            emitCaseOrDefaultClauseStatements(node, node.statements);
+            const pos = emitTokenWithComment(SyntaxKind.DefaultKeyword, node.pos, writeKeyword, node);
+            emitCaseOrDefaultClauseStatements(node, node.statements, pos);
         }
 
-        function emitCaseOrDefaultClauseStatements(parentNode: Node, statements: NodeArray<Statement>) {
+        function emitCaseOrDefaultClauseStatements(parentNode: Node, statements: NodeArray<Statement>, colonPos: number) {
+            // This also handles the colon token from the preceeding clause, as weather comments must be emitted for it depends on
```
Comment (rbuckton) [inline src/compiler/emitter.ts]: nit: This comment is hard to read and has misspellings ('whether' not 'weather'), can you rephrase it?

Comment (rbuckton) [reply src/compiler/emitter.ts]: You can probably just remove the comment entirely if you rename the function to `emitCaseOrDefaultClauseRest` (as it no longer would imply only emitting the statements).

**Comment by rbuckton** [comment]
Can you address the merge conflict?
